Build error on main under Windows w/ CUDA 12.2

### Description

I can successfully build from the v12 branch. When I try to build from main, the build fails.

I've attached a complete output from the build as a file. [full_build_log.txt](https://github.com/cupy/cupy/files/15407148/full_build_log.txt)

Here are some excerpts that seem potentially relevant. 

I get many instances of the following warning, which appears possibly related to a subsequent error:

```
  C:\Users\phrutschman\Documents\cupy build\cupy\cupy\cuda\cupy_cub.cu(77): warning #20013-D: calling a constexpr __host__ function("infinity") from a __host___device__ function("infinity") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
            return thrust::complex<double>(std::numeric_limits<double>::infinity(), std::numeric_limits<double>::infinity());
```

Followed by this error
```
  C:\Users\phrutschman\Documents\cupy build\cupy\cupy\cuda\cupy_cub.cu(86): error C3615: constexpr function 'std::numeric_limits<__half>::infinity' cannot rest in a constant expression
  C:\Users\phrutschman\Documents\cupy build\cupy\cupy\cuda\cupy_cub.cu(88): note: failure was caused by a statement or an expression that is not valid in a cotexpr context
  tmpxft_00002704_00000000-10_cupy_thrust.compute_90.cudafe1.cpp
```

followed by this error
```
  Traceback (most recent call last):
    File "C:\Users\phrutschman\Documents\cupy build\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
      main()
    File "C:\Users\phrutschman\Documents\cupy build\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\phrutschman\Documents\cupy build\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 251, in build_wheel
      return _build_backend().build_wheel(wheel_directory, config_settings,
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\build_meta.py", line 410, in build_wheel
      return self._build_with_temp_dir(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\build_meta.py", line 395, in _build_with_temp_di
      self.run_setup()
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\build_meta.py", line 487, in run_setup
      super().run_setup(setup_script=setup_script)
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\build_meta.py", line 311, in run_setup
      exec(code, locals())
    File "<string>", line 127, in <module>
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\__init__.py", line 103, in setup
      return distutils.core.setup(**attrs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\_distutils\core.py", line 184, in setup
      return run_commands(dist)
             ^^^^^^^^^^^^^^^^^^
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\_distutils\core.py", line 200, in run_commands
      dist.run_commands()
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 969, in run_commands
      self.run_command(cmd)
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\dist.py", line 968, in run_command
      super().run_command(command)
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
      cmd_obj.run()
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\wheel\bdist_wheel.py", line 368, in run
      self.run_command("build")
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\_distutils\cmd.py", line 316, in run_command
      self.distribution.run_command(command)
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\dist.py", line 968, in run_command
      super().run_command(command)
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
      cmd_obj.run()
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\_distutils\command\build.py", line 132, in run
      self.run_command(cmd_name)
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\_distutils\cmd.py", line 316, in run_command
      self.distribution.run_command(command)
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\dist.py", line 968, in run_command
      super().run_command(command)
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
      cmd_obj.run()
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\command\build_ext.py", line 91, in run
      _build_ext.run(self)
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\normal\Lib\site-packages\Cython\Distutils\old_build_ext.py", line 186, in run
      _build_ext.build_ext.run(self)
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\_distutils\command\build_ext.py", line 359, in r
      self.build_extensions()
    File "C:\Users\phrutschman\Documents\cupy build\cupy\install\cupy_builder\_command.py", line 176, in build_extensions
      super().build_extensions()
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\normal\Lib\site-packages\Cython\Distutils\old_build_ext.py", line 195, in build_extenons
      _build_ext.build_ext.build_extensions(self)
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\_distutils\command\build_ext.py", line 477, in bld_extensions
      self._build_extensions_parallel()
    File "C:\Users\phrutschman\AppData\Local\Temp\pip-build-env-vt139mgo\overlay\Lib\site-packages\setuptools\_distutils\command\build_ext.py", line 500, in _ild_extensions_parallel
      fut.result()
    File "C:\Users\phrutschman\AppData\Local\Programs\Python\Python311\Lib\concurrent\futures\_base.py", line 456, in result
      return self.__get_result()
             ^^^^^^^^^^^^^^^^^^^
    File "C:\Users\phrutschman\AppData\Local\Programs\Python\Python311\Lib\concurrent\futures\_base.py", line 401, in __get_result
      raise self._exception
    File "C:\Users\phrutschman\AppData\Local\Programs\Python\Python311\Lib\concurrent\futures\thread.py", line 58, in run
      result = self.fn(*self.args, **self.kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\phrutschman\Documents\cupy build\cupy\install\cupy_builder\_command.py", line 200, in build_extension
      sources_cpp, extra_objects = compile_device_code(ctx, ext)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "C:\Users\phrutschman\Documents\cupy build\cupy\install\cupy_builder\_command.py", line 64, in compile_device_code
      compiler.compile(obj, src, ext)
    File "C:\Users\phrutschman\Documents\cupy build\cupy\install\cupy_builder\_compiler.py", line 256, in compile
      self.spawn(compiler_so + cc_args + [src, '-o', obj] + postargs)
    File "C:\Users\phrutschman\Documents\cupy build\cupy\install\cupy_builder\_compiler.py", line 180, in spawn
      subprocess.check_call(commands)
    File "C:\Users\phrutschman\AppData\Local\Programs\Python\Python311\Lib\subprocess.py", line 413, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['C:\\Program Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v12.2\\bin/nvcc.exe', '-D_FORCE_INLINES=1', '-DCUPY_CACHE_K=3a0dbb91e283c0fe3ada96761f63f2b8c0b67794', '-DCUPY_CUB_VERSION_CODE=200200', '-DCUPY_JITIFY_VERSION_CODE=1a0ca0e', '-IC:\\Users\\phrutschman\\Documents\\cupyuild\\cupy\\cupy/_core/include\\cupy/_cccl/libcudacxx', '-IC:\\Users\\phrutschman\\Documents\\cupy build\\cupy\\cupy/_core/include\\cupy/_cccl/thrust', '-IC:\sers\\phrutschman\\Documents\\cupy build\\cupy\\cupy/_core/include\\cupy/_cccl/cub', '-IC:\\Users\\phrutschman\\Documents\\cupy build\\cupy\\cupy/_core/includ, '-IC:\\Program Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v12.2\\include', '-IC:\\Program Files\\NVIDIA Corporation\\Nsight Systems 2023.2.3\\target-windows64\\nvtx\\include', '-c', 'cupy/cuda/cupy_cub.cu', '-o', 'build/temp.device_objects/cupy/cuda/cupy_cub.cu.obj', '--generate-code=arch=compute_50,code=sm_50', -generate-code=arch=compute_52,code=sm_52', '--generate-code=arch=compute_60,code=sm_60', '--generate-code=arch=compute_61,code=sm_61', '--generate-code=arch=mpute_70,code=sm_70', '--generate-code=arch=compute_75,code=sm_75', '--generate-code=arch=compute_80,code=sm_80', '--generate-code=arch=compute_86,code=sm_86''--generate-code=arch=compute_89,code=sm_89', '--generate-code=arch=compute_90,code=sm_90', '--generate-code=arch=compute_90,code=compute_90', '-Xfatbin=-compss-all', '-O2', '-allow-unsupported-compiler', '-Xcompiler', '/MD', '-D_USE_MATH_DEFINES', '--std=c++17', '-t2', '--compiler-bindir', 'C:\\Program Files (x86)Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.37.32822\\bin\\HostX64\\x64']' returned non-zero exit status 4294967295.
  [end of output]
```





### To Reproduce

Enable symlinks in windows + git

```
git clone https://github.com/cupy/cupy.git
cd cupy
pip install .
```


### Installation

Built from GitHub source

### Environment

```
n/a cupy is not yet built at the time of the error
```


### Additional Information

Versions:
Windows: 10.0.19045
MSVC: 14.37.32822
CUDA: 12.2

--- a/cupy/cuda/cupy_cub.cu
+++ b/cupy/cuda/cupy_cub.cu
@@ -84,7 +84,7 @@ template <>
 class numeric_limits<__half> {
   public:
     static __host__ __device__ constexpr __half infinity() noexcept {
-        constexpr unsigned short inf_half = 0x7C00U;
+        unsigned short inf_half = 0x7C00U;
         #if (defined(_MSC_VER) && _MSC_VER >= 1920)
         // WAR:
         // - we want a constexpr here, but reinterpret_cast cannot be used

--- a/cupy/cuda/cupy_cub.cu
+++ b/cupy/cuda/cupy_cub.cu
@@ -85,8 +85,11 @@ class numeric_limits<__half> {
   public:
     static __host__ __device__ constexpr __half infinity() noexcept {
         unsigned short inf_half = 0x7C00U;
-        __half inf_value = *reinterpret_cast<__half*>(&inf_half);
-        return inf_value;
+        // WAR:
+        // - we want a constexpr here, but reinterpret_cast cannot be used
+        // - we want to use std::bit_cast, but it requires C++20 which is too new
+        // - we use the compiler builtin, fortunately both gcc and msvc have it
+        return __builtin_bit_cast(__half, inf_half);
     }
 
     static constexpr bool has_infinity = true;

--- a/cupy/cuda/cupy_cub.cu
+++ b/cupy/cuda/cupy_cub.cu
@@ -84,12 +84,16 @@ template <>
 class numeric_limits<__half> {
   public:
     static __host__ __device__ constexpr __half infinity() noexcept {
-        unsigned short inf_half = 0x7C00U;
+        constexpr unsigned short inf_half = 0x7C00U;
+        #if (defined(_MSC_VER) && _MSC_VER >= 1920)
         // WAR:
         // - we want a constexpr here, but reinterpret_cast cannot be used
         // - we want to use std::bit_cast, but it requires C++20 which is too new
         // - we use the compiler builtin, fortunately both gcc and msvc have it
         return __builtin_bit_cast(__half, inf_half);
+        #else
+        return *reinterpret_cast<__half*>(&inf_half);
+        #endif
     }
 
     static constexpr bool has_infinity = true;

